### ISSUE TITLE
redirectDomain and httpsOnly

### ISSUE THREAD

**Issue by rufinus** [vfarcic/docker-flow-proxy#401]
When using this labels in a stack config:
```
            labels:
                - com.df.serviceDomain=rainbowbridge.rip
                - com.df.redirectFromDomain=www.rainbowbridge.rip,www.regenbogenbruecke.at,www.regenbogenbruecke.rip,regenbogenbruecke.at,regenbogenbruecke.rip
                - com.df.port=80
                - com.df.notify=true
                - com.df.distribute=true
                - com.df.httpsOnly=true
                - com.df.httpsRedirectCode=301
                - com.df.srcPort=443
                - com.df.compressionAlgo=gzip
                - com.df.compressionType=text/css text/html text/javascript application/javascript text/plain text/xml application/json
```
i get this haproxy config:

```
    compression algo gzip                                                                                                 
    compression type text/css text/html text/javascript application/javascript text/plain text/xml application/json       
    acl url_rainbowbridge_apache80_0 path_beg /                                                                           
    acl srcPort_rainbowbridge_apache443 dst_port 443                                                                      
    acl domain_rainbowbridge_apache80_0 hdr_beg(host) -i rainbowbridge.rip                                                
    http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i www.rainbowbridge.rip }          
    http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.at }       
    http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.rip }      
    http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.at }           
    http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.rip }          
    use_backend rainbowbridge_apache-be80_0 if url_rainbowbridge_apache80_0 domain_rainbowbridge_apache80_0 srcPort_rainbowbridge_apache443
                                                                                                                                                                                                                                                                                                                                                      
backend rainbowbridge_apache-be80_0                                                                                                        
    mode http                                                                                                                              
    http-request redirect scheme https code 301 if !{ ssl_fc }                                                                             
    http-request add-header X-Forwarded-Proto https if { ssl_fc } 
```
Whats not working:
- `http://rainbowbridge.rip/` does not redirect to https:// (and shows 503)
- `https://<redirectDomain>` redirects to `http://rainbobridge.rip` (instead of https://rainbowbridge.rip)

https://rainbowbridge.rip/ does work as expected.

Where did i make the misconfiguration?

Other question - do the http config vars work with reqMode=sni  (docu explicit mention http mode only)

**Comment by vfarcic** [comment]
The second issue (redirect does not maintain the protocol (http/https) should be fixed. I made a temporary image `vfarcic/docker-flow-proxy:issue-401` for you to try it out. If it works as expected,  I'll promote it to an official release and, from there on, we can move to the first issue. Does that sound OK?

As for SNI, most http config vars do not work with it. SNI uses TCP with a headerless protocol. As such, many of the HTTP features will not work. Adding @faisyl to the thread. He's the author behind SNI support and will probably be able to provide a more detailed answer.

**Comment by rufinus** [comment]
@vfarcic wow, didnt expected a response that fast. thank you.

I tried issue-401 version:
`[ALERT] 359/125756 (71) : parsing [/cfg/haproxy.cfg:65] : error detected in frontend 'services' while parsing 'http-request redirect' rule : error in condition: no such ACL : '}'.
[ALERT] 359/125756 (71) : Error(s) found in configuration file : /cfg/haproxy.cfg`

proxy config now looks like:
```
    compression algo gzip
    compression type text/css text/html text/javascript application/javascript text/plain text/xml application/json
    acl url_rainbowbridge_apache80_0 path_beg /
    acl srcPort_rainbowbridge_apache443 dst_port 443
    acl domain_rainbowbridge_apache80_0 hdr_beg(host) -i rainbowbridge.rip
	http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i www.rainbowbridge.rip !{ ssl_fc } }
	http-request redirect code 301 prefix https://rainbowbridge.rip if { hdr_beg(host) -i www.rainbowbridge.rip { ssl_fc } }
	http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.at !{ ssl_fc } }
	http-request redirect code 301 prefix https://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.at { ssl_fc } }
	http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.rip !{ ssl_fc } }
	http-request redirect code 301 prefix https://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.rip { ssl_fc } }
	http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.at !{ ssl_fc } }
	http-request redirect code 301 prefix https://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.at { ssl_fc } }
	http-request redirect code 301 prefix http://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.rip !{ ssl_fc } }
	http-request redirect code 301 prefix https://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.rip { ssl_fc } }
    use_backend rainbowbridge_apache-be80_0 if url_rainbowbridge_apache80_0 domain_rainbowbridge_apache80_0 srcPort_rainbowbridge_apache443

```

**Comment by vfarcic** [comment]
Sorry about that. I don't have SSL cert available for testing purposes so I made a "blind" change.

I found the problem with the first issue. You should remove `com.df.srcPort=443`. From there on, http should be redirected to https.

As for the second problem, I made another "blind" change. Can you pull the same tag `issue-401` again and try it out?

**Comment by rufinus** [comment]
thank you @vfarcic its working now.

haproxy config is now:
```
    compression algo gzip
    compression type text/css text/html text/javascript application/javascript text/plain text/xml application/json
    acl url_rainbowbridge_apache80_0 path_beg /
    acl domain_rainbowbridge_apache80_0 hdr_beg(host) -i rainbowbridge.rip
    http-request redirect code 301 prefix %[var(req.scheme)]://rainbowbridge.rip if { hdr_beg(host) -i www.rainbowbridge.rip }
    http-request redirect code 301 prefix %[var(req.scheme)]://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.at }
    http-request redirect code 301 prefix %[var(req.scheme)]://rainbowbridge.rip if { hdr_beg(host) -i www.regenbogenbruecke.rip }
    http-request redirect code 301 prefix %[var(req.scheme)]://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.at }
    http-request redirect code 301 prefix %[var(req.scheme)]://rainbowbridge.rip if { hdr_beg(host) -i regenbogenbruecke.rip }
    use_backend rainbowbridge_apache-be80_0 if url_rainbowbridge_apache80_0 domain_rainbowbridge_apache80_0
```
this will work for now - but as soon as i will have a second https enabled page on this stack i will run into the SNI problem i guess. according to my research this is just not working with haproxy (or not very well) - the only solution i found is with an apache in front of haproxy for the ssl termination.

**Comment by faisyl** [comment]
SNI routing works for HAProxy, though I'm not sure you can mix both SSL
termination and SNI routing in the same front-end port. See the
documentation on how to use SNI routing and feel free to mail me if you
need help with that.

**Comment by faisyl** [comment]
If you are okay terminating SSL on HAProxy, then you can get by without SNI and let HAProxy present the correct certificate for the domain. This should work as long as the server certificates' Subject name or SAN matches the SNI in the connection. Although HAProxy will then use the SNI field to determine the certificate to present, you needn’t configure SNI based routing to handle the backend routing.

Thanks!
~faisal

**Comment by vfarcic** [comment]
@rufinus What @faisyl explained is correct. If your problem is with certificates for multiple domains, that should work.

I'll close this issue since the original issue is fixed. Please let me know if you have problem with multi-domain setup.

**Comment by vfarcic** [comment]
The fix will be available in the tag `vfarcic/docker-flow-proxy:17.12.26-93`. It's building while I write this...

**Comment by rufinus** [comment]
thanks @vfarcic & @faisyl 
SNI termination on haproxy is perfectly fine. i dont see much reason in having the ssl overhead between backend and haproxy.
